cirq-google/cirq_google/engine/simulated_local_processor.py

```diff
@@ -0,0 +1,319 @@
+    elif isinstance(union_time, datetime.date):
+        return int(datetime.datetime.combine(union_time, datetime.datetime.min.time()).timestamp())
+    return None
+
+
+class SimulatedLocalProcessor(AbstractLocalProcessor):
+    """A processor backed by a sampler and device.
+
+    Intended for local simulation testing, this processor will
+    create a VAlidationSampler that will validate requests based on
+    the provided device and an additional Callable (that can verify
+    serialization constraints, for instance).  Jobs will then be
+    executed using the provided sampler.
+
+    This class also supports a list of calibration metrics that are
+    stored in-memory to replicate Quantum Engine calibration metrics.
+
+    This class can be used as a local emulator for the Quantum Engine
+    API or for testing or mocking.
+
+    Attributes:
+        sammpler: A `cirq.Sampler` that can execute the quantum jobs.
+        device: An optional device, for validation of qubit connectivity.
+        validator: A Callable that can validate additional characteristics
+            beyond the device, such as serialization, repetition limits, etc.
+        calibrations: A dictionary of calibration metrics keyed by epoch seconds
+            that can be returned by the processor.
+    """
+
+    def __init__(
+        self,
+        *args,
+        sampler: cirq.Sampler = cirq.Simulator(),
+        device: cirq.Device = cirq.UNCONSTRAINED_DEVICE,
+        validator: validating_sampler.VALIDATOR_TYPE = None,
+        simulation_type: LocalSimulationType = LocalSimulationType.SYNCHRONOUS,
+        calibrations: Optional[Dict[int, calibration.Calibration]] = None,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self._calibrations = calibrations or {}
+        self._device = device
+        self._simulation_type = simulation_type
+        self._validator = validator
+        self._sampler = validating_sampler.ValidatingSampler(
+            device=self._device, validator=self._validator, sampler=sampler
+        )
+        self._programs: Dict[str, AbstractLocalProgram] = {}
+
+    def remove_program(self, program_id: str):
+        """Remove reference to a child program."""
+        if program_id in self._programs:
+            del self._programs[program_id]
+
+    def get_calibration(self, calibration_timestamp_seconds: int) -> calibration.Calibration:
+        return self._calibrations[calibration_timestamp_seconds]
+
+    def get_latest_calibration(self, timestamp: int) -> Optional[calibration.Calibration]:
+        latest = None
+        current_calibration = None
+        for calibration_seconds in self._calibrations:
+            if calibration_seconds <= timestamp and (
+                latest is None or latest < calibration_seconds
+            ):
+                latest = calibration_seconds
+                current_calibration = self._calibrations[latest]
+        return current_calibration
```

Comment:
[Optional] There's probably a more concise/pythonic way of doing this with `max(..., key=...)`. Leaving it up to you whether to switch to that form.